cpp/ql/src/semmle/code/cpp/models/implementations/Strtok.qll

```diff
@@ -0,0 +1,49 @@
+/**
+ * Provides implementation classes modeling `strtok` and various similar
+ * functions. See `semmle.code.cpp.models.Models` for usage information.
+ */
+
+import semmle.code.cpp.Function
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * The standard function `strtok` and its assorted variants
+ */
+private class Strtok extends ArrayFunction, AliasFunction, TaintFunction, SideEffectFunction {
+  Strtok() {
+    this.hasGlobalOrStdName("strtok") or
+    this.hasGlobalName(["strtok_r", "_strtok_l", "wcstok", "_wcstok_l", "_mbstok", "_mbstok_l"])
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) { bufParam = [0, 1] }
+
+  override predicate hasArrayInput(int bufParam) { bufParam = [0, 1] }
+
+  override predicate hasArrayOutput(int bufParam) { bufParam = 0 }
+
+  override predicate parameterNeverEscapes(int index) { index = 1 }
+
+  override predicate parameterEscapesOnlyViaReturn(int index) { index = 0 }
+
+  override predicate parameterIsAlwaysReturned(int index) { none() }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and output.isReturnValue()
+  }
+
+  override predicate hasOnlySpecificReadSideEffects() { any() }
+
+  override predicate hasOnlySpecificWriteSideEffects() { any() }
```

Comment:
I'm not sure these are accurate.  http://www.cplusplus.com/reference/cstring/strtok/ says: "The point where the last token was found is kept internally by the function to be used on the next call (particular library implementations are not required to avoid data races)."  This suggests some internal state is being read and written.

The QLDoc for these methods does say: "reads from memory that cannot be visible to the caller (for example a buffer inside an I/O library) are not modeled here.", but I'm not sure that applies, because the write to this memory does affect the result we're given in a follow-up call.

Reply:
Hm, good point. I guess you can observe these memory effects. I'll remove these two overrides.

Reply:
These methods are about whether a call to this function should have the generic side-effect write to `AllAliasedMemory`. I believe that's not the case for `strtok`, since pointers available to other functions should never alias with that buffer.

Reply:
I guess we can't both specify that:
- `strtok` has some hidden state, and
- that this state does not alias with anything outside `strtok`

right?

So @geoffw0 argues that we should go for the sound option and say that this hidden state _can_ alias with the rest of `AllAliasedMemory`.